aeneas/src/ir/SsaNormalizer.v3

```diff
@@ -1370,18 +1369,53 @@ class SsaRaNormalizer extends SsaRebuilder {
 
 		var result = Array<SsaInstr>.new(vn.size);
 		for (i < result.length) result[i] = newGraph.nullConst(vn.at(i));
-		if (!vn.isTagless()) result[vn.tagIndex()] = newGraph.intConst(vn.tagValue);
 		
+		if (vn.hasExplicitTag()) {
+			if (vn.tag.isPacked()) {
+				var tagIdx = vn.tag.indexes[0];
+				result[tagIdx] = genSetInterval(result[tagIdx], newGraph.intConst(vn.tagValue), vn.tag.intervals[0], vn.tag.tn.newType, IntRepType.!(vn.at(tagIdx)));
+			} else {
+				result[vn.tagIndex()] = newGraph.intConst(vn.tagValue);
+			}
+		}
+
 		for (i < vn.fields.length) {
 			var f = vn.fields[i];
 			var fieldRanges = vn.fieldRanges[i], os = fieldRanges.0;
 			for (j < f.indexes.length) {
 				var idx = f.indexes[j];
-				result[idx] = genVariantScalarView(f.tn.at(j), vn.at(idx), ai_inputs[os + j]);
+				if (!f.isPacked()) result[idx] = genVariantScalarView(f.tn.at(j), vn.at(idx), ai_inputs[os + j]);
+				else result[idx] = genSetInterval(result[idx], ai_inputs[os + j], f.intervals[j], f.tn.at(j), IntRepType.!(vn.at(idx)));
 			}
 		}
 		return result;
 	}
+	def genExtractInterval(scalar: SsaInstr, interval: Interval, ft: IntRepType, tt: Type) -> SsaInstr {
+		if (interval.start > 0) scalar = curBlock.addApplyF(ft.opShr(), [scalar, newGraph.intConst(interval.start)], Fact.O_PURE | Fact.O_NO_SHIFT_CHECK);
```

Comment:
`Fact.O_PURE | Fact.O_NO_SHIFT_CHECK` occurs several times; might be worth factoring out to `Facts`.